### PR TITLE
Chore: add source map to static build

### PR DESCRIPTION
**Pull request by leangseu-edx**
**TL;DR -**
Improvement to developing javascript using webpack build:
- [x] include source map
- [x] has distinct name on production build
- [x] `npm run build` to build production
- [x] `npm run dev` to build development
- [x] `npm run start` to start development server

Bonus:
- [x] Hot reload in dev build. **CMS** does not hot reload yet.

JIRA: [educator 5480](https://openedx.atlassian.net/browse/EDUCATOR-5480)

**What changed?**

- `make static` should build static javascript that include sourceMap
- breakpoint within prod build should work
- `npm run start` should 

**Developer Checklist**

- [x] Reviewed the [release process](https://github.com/edx/edx-ora2/blob/master/.github/release_process.md)
- [x] Translations and JS/SASS compiled
- [x] Bumped version number in [setup.py](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/setup.py#L39) and [package.json](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/package.json#L3)

**Testing Instructions**

** Hot Reload **
- from local directory (not in lms shell)
- `npm run start`
- from devstack directory (not in lms shell)
- `make lms-restart studio-restart` This is only require when you `load_static.py` to reload its static method.
  - **NOTE**: cms does not support hot reload at the moment
- modify `javascript` or add `debugger` within lms page that has `xblock-ora`. It should hot reload the `xblock`.
  - **NOTE**: lms use iframe while cms is not. One of the reason hot reload not working properly.

**Reviewer Checklist**
- [x] `make static`
- [x] `npm run dev`
- [x] `npm run start`

NOTE:
- Each command require `make lms-restart studio-restart`
- Source map should be working
- `Debugger` should only work on development build

Collectively, these should be completed by reviewers of this PR:

- [x] I've done a visual code review
- [x] I've tested the new functionality

FYI: @edx/masters-devs-gta

### CODE REVIEW

**Comment by leangseu-edx** [comment]
Currently, check translation is including `static/dist/*`. I was thinking of adding that folder into gitignore but it does not seems to pass travis CI.

**Comment by leangseu-edx** [comment]
**NOTE**: File to emphasis is `load_static.py`. Is there a prefer way to do it? Is there a prefer place to put the file.

File: openassessment/xblock/load_static.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    def reload_manifest():
+        """
+        Reload from manifest file
+        """
+        json_data = resource_string(__name__, 'static/dist/manifest.json').decode("utf8")
+        LoadStatic._manifest = json.loads(json_data)
+        if LoadStatic._manifest['is_dev_server']:
+            host = ''
+        elif hasattr(settings, 'LMS_ROOT_URL'):
+            host = settings.LMS_ROOT_URL
```
Comment (leangseu-edx) [inline openassessment/xblock/load_static.py:31]: Assuming this variable is being use correctly. This is one of the reason the older code was using `add_javascript` instead of `add_javascript_url`. CMS could not get the script because there isn't iframe for proxy to CND.

**Review by jnlapierre — APPROVED** [review]
LGTM

File: webpack.prod.config.js [hunk 2]
```diff
@@ -75,26 +74,30 @@ Object.assign(config, {
...
+      }
+    })
   ],
 });
 
-config.resolve.modules = ['node_modules'].concat(
-  path.resolve(__dirname, 'openassessment/xblock/static/js/src'),
-);
+config.resolve.modules = ['node_modules', path.resolve(__dirname, 'openassessment/xblock/static/js/src')]
+config.output.path = path.resolve(process.cwd(), 'openassessment/xblock/static/dist')
```
Comment (muselesscreator) [inline webpack.prod.config.js:101]: eslint errors in this file (missing commas/semicolons, and wrong quotes). not all of these complaints are new w/ this change.

File: webpack.dev.config.js [hunk 3]
```diff
@@ -88,13 +89,19 @@ Object.assign(config, {
...
+      writeToFileEmit: true
     }),
+    new webpack.HotModuleReplacementPlugin(),
   ],
 });
 
-config.resolve.modules = ['node_modules'].concat(
-  path.resolve(__dirname, 'openassessment/xblock/static/js/src'),
-);
+config.resolve.modules = ['node_modules', path.resolve(__dirname, 'openassessment/xblock/static/js/src')]
```
Comment (muselesscreator) [inline webpack.dev.config.js:105]: eslint errors in this file (missing commas/semicolons, and wrong quotes). not all of these complaints are new w/ this change.

Comment (muselesscreator) [inline webpack.dev.config.js:105]: eslint errors in this file (missing commas/semicolons, and wrong quotes). not all of these complaints are new w/ this change.

File: openassessment/xblock/load_static.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        LoadStatic._base_url = host + LoadStatic._manifest['base_url']
+
+    @staticmethod
+    def get_url(key):
+        """
+        get url from key
+        """
+        if not LoadStatic._base_url:
+            LoadStatic.reload_manifest()
+        return LoadStatic._base_url + LoadStatic._manifest[key]
```
Comment (leangseu-edx) [inline openassessment/xblock/load_static.py:44]: Hopefully, this is okay in `stage/prod` allow CND script from ORA. If not, we might need additional configuration on the server.

File: docs/developers_guide.rst [hunk 5]
```diff
@@ -85,6 +85,16 @@ This is required if there were any JS changes:
 - from local directory (not in lms shell)
 - ``npm run build``
 
+Hot reload JS
```
Comment (nsprenkle) [inline docs/developers_guide.rst:88]: I had a fair bit of trouble finding where verison of the files were in the browser developer tools. Bonus points for adding docs on how to do that. Also notes around port collisions, since my dev server defaulted to `8080` which docker already uses.

File: openassessment/xblock/static/dist/manifest.json [hunk 6]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (nsprenkle) [inline openassessment/xblock/static/dist/manifest.json:1]: Should we be including this manifest file or is this just a developer file we should `.gitignore` out of GitHub?

Comment (leangseu-edx) [reply openassessment/xblock/static/dist/manifest.json:1]: I agree that it is generated. However, it is entangle with javascript translation. For now,  it is better just to leave it in.

**Comment by leangseu-edx** [comment]
> Works well for me 👍
> 
> Wanted to check about the renaming of dist files with hashes, is it best practice to include those with the dist files?

It is part of the AC and yes it is a good practice. I also saw it included in fedx scripts. 

> Even though the generated build works, the file names it uses is problematic. It doesn't allow to leverage client level cache efficiently as there's no way tell whether or not a file has changed. Cache invalidation can be achieved by including a hash to the filenames.

File: .env.development [hunk 7]
```diff
@@ -3,3 +3,4 @@ NODE_ENV=development
 BABEL_ENV=development
 NODE_PATH=./openassessment/xblock/static/js/src
 BROWSER=none
+PORT=8080 # port for webpack dev server
```
Comment (nsprenkle) [inline .env.development:6]: Add a newline to file.

Don't use 8080 if possible, lots of other dev stuff defaults to this port.

File: docs/developers_guide.rst [hunk 8]
```diff
@@ -119,6 +120,11 @@ The simplest way to debug ORA2 code is with PDB - Python's built in debugger.
 One caveat: Since ORA2 has code that executes either in studio or lms context, one must be attached to the corret shell 
 in order for the breakpoints to be hit.
 
+Debugging JavaScript
+------------------
+Adding ``debugger`` to the JavaScript should work for ``development build`` and ``development server``.
```
Comment (nsprenkle) [inline docs/developers_guide.rst:125]: I'd even expand to say, look for this folder path in the dev tools. It took me a while to find on my own...

**Comment by nsprenkle** [comment]
Does this change anything about our build/commit process? e.g. do I need to be committing new files that weren't there before? If anything has changed it should be added to release/PR checklists.

File: .gitignore [hunk 9]
```diff
@@ -88,6 +88,4 @@ edx-ora2/
 private.py
 
 # pyenv
-.python-version
-
-!openassessment/xblock/static/dist/
+.python-version
```
Comment (nsprenkle) [inline .gitignore:91]: Add missing newline

**Comment by jansenk** [comment]
When I tried to run your branch, it complained about a missing `openassessment/xblock/static/dist/manifest.json`. Running make install generated the file. Should it be included in the repo? I also noticed that `openassessment/xblock/static/dist/openassessment-lms.c98e88f0855153453b82.js` became `openassessment/xblock/static/dist/openassessment-lms.f4c1630142b43804ec97.js`. Should that be checked in?

File: docs/developers_guide.rst (hunk 8 above)
Comment (jansenk) [reply docs/developers_guide.rst:125]: Could you add a comment with where they are? I can't find them

File: .env.development (hunk 7 above)
Comment (leangseu-edx) [reply .env.development]: I changed it to 18001...

File: docs/developers_guide.rst (hunk 8 above)
Comment (leangseu-edx) [reply docs/developers_guide.rst:125]: Where would you suggest me changing? I am not sure where else to put it.

**Comment by leangseu-edx** [comment]
> When I tried to run your branch, it complained about a missing `openassessment/xblock/static/dist/manifest.json`. Running make install generated the file. Should it be included in the repo? I also noticed that `openassessment/xblock/static/dist/openassessment-lms.c98e88f0855153453b82.js` became `openassessment/xblock/static/dist/openassessment-lms.f4c1630142b43804ec97.js`. Should that be checked in?

Every time `javascript` change, the chunk hash will get change. I was trying to remove all of the files within `dist` folder and include it in `gitignore`. They are generated. However, translation process require javascript files and it has to be the latest. I do not want to make a decision call for changing any of that at the moment.

File: .env.development [hunk 10]
```diff
@@ -0,0 +1,6 @@
+RC_ENV=development
+NODE_ENV=development
+BABEL_ENV=development
+NODE_PATH=./openassessment/xblock/static/js/src
+BROWSER=none
+PORT=18001 # port for webpack dev server
```
Comment (nsprenkle) [inline .env.development:6]: Add missing newline

File: docs/developers_guide.rst [hunk 11]
```diff
@@ -109,6 +120,11 @@ The simplest way to debug ORA2 code is with PDB - Python's built in debugger.
...
```
Comment (nsprenkle) [inline docs/developers_guide.rst:125]:
```suggestion
Adding ``debugger`` to the JavaScript should work for ``development build`` and ``development server``.
```

File: docs/developers_guide.rst [hunk 12]
```diff
@@ -109,6 +120,11 @@ The simplest way to debug ORA2 code is with PDB - Python's built in debugger.
...
+It should not work in ``production build`` however, adding breakpoint in mapped files for debugging should work on both build.
+
```
Comment (nsprenkle) [inline docs/developers_guide.rst:127]: Final proposed wording change (with @leangseu-edx to fill in locations from browser debug tools)

```suggestion
Debugging JavaScript
------------------
For debugging JS in Devstack, first follow the instructions for "Hot Reload JS". This enables source maps and allows for placing breakpoints in source-mapped files from the browser dev tools.

 Locate code by browsing to <Leangseu, add path here> and breakpoints should toggle with hot-reloading.

```

**Review by nsprenkle — APPROVED** [review]
Cannot stress how much of a developer win this is 💯 

A final proposed wording change to make it more obvious where to find source-mapped files (since it took me like 10 minutes and I want it to be more obvious to other devs).
